pytest failure when running out-of-tree plugin against 2.25.0a0 - macOS wheels include single architecture `native_engine.so` files

Problem: Running pytest under Pants v2.25.0a0 is failing on my macOS/x86-64 system. The project under test is a out-of-tree Pants plugin and the test thus imports from the `pantsbuild.pants` distribution, hence why native_engine.so is being loaded in a test. (See the specific error below.)

The `native_engine.so` being imported appears to be for arm64 and not for x86-64:

```
HOST:~/.cache/pants/named_caches/pex_root/venvs/1/s/79b3222f/venv/lib/python3.11/site-packages/pants/engine/internals% file native_engine.so
native_engine.so: Mach-O 64-bit dynamically linked shared library arm64
```

We seem to be publishing universal2 macOS wheels instead of the previous separate architecture-specific wheels.
[2.24.1](https://github.com/pantsbuild/pants/releases/tag/release_2.24.1):
- pantsbuild.pants-2.24.1-cp39-cp39-macosx_10_15_x86_64.whl
- pantsbuild.pants-2.24.1-cp39-cp39-macosx_11_0_arm64.whl

[2.25.0a0](https://github.com/pantsbuild/pants/releases/tag/release_2.25.0a0):
- pantsbuild.pants-2.25.0a0-cp311-cp311-macosx_10_12_universal2.whl
- pantsbuild.pants-2.25.0a0-cp311-cp311-macosx_11_0_universal2.whl

Unpacking the wheels, we see that the `.so` files in the root of the wheel are multi-architecture. Yet, the `native_engine.so` in the actual package source directory is one architecture only. (x86-64 for the 10.12 wheel and arm64 for the 11.0 wheel.)

```
./unpacked-wheel-11.0-universal2/pants/engine/internals/native_engine.so:  Mach-O 64-bit dynamically linked shared library arm64

./unpacked-wheel-11.0-universal2/native_engine.cpython-311-darwin.so:      Mach-O universal binary with 2 architectures: [x86_64:Mach-O 64-bit bundle x86_64] [arm64:Mach-O 64-bit bundle arm64]
./unpacked-wheel-11.0-universal2/native_engine.cpython-311-darwin.so (for architecture x86_64):	Mach-O 64-bit bundle x86_64
./unpacked-wheel-11.0-universal2/native_engine.cpython-311-darwin.so (for architecture arm64):	Mach-O 64-bit bundle arm64

./unpacked-wheel-10.12-universal2/pants/engine/internals/native_engine.so: Mach-O 64-bit dynamically linked shared library x86_64

./unpacked-wheel-10.12-universal2/native_engine.cpython-311-darwin.so:     Mach-O universal binary with 2 architectures: [x86_64:Mach-O 64-bit bundle x86_64] [arm64:Mach-O 64-bit bundle arm64]
./unpacked-wheel-10.12-universal2/native_engine.cpython-311-darwin.so (for architecture x86_64):	Mach-O 64-bit bundle x86_64
./unpacked-wheel-10.12-universal2/native_engine.cpython-311-darwin.so (for architecture arm64):	Mach-O 64-bit bundle arm64
```


<hr/>

The full error that I saw when trying to run my plugin's test was:

```
_ ERROR collecting src/python/shoalsoft/FOO/BAR_test.py _
ImportError while importing test module 'src/python/shoalsoft/FOO/BAR_test.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/Users/ME/.cache/pants/named_caches/pex_root/venvs/1/s/79b3222f/venv/lib/python3.11/site-packages/_pytest/python.py:493: in importtestmodule
    mod = import_path(
/Users/ME/.cache/pants/named_caches/pex_root/venvs/1/s/79b3222f/venv/lib/python3.11/site-packages/_pytest/pathlib.py:587: in import_path
    importlib.import_module(module_name)
/Users/ME/.pyenv/versions/3.11.11/lib/python3.11/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
<frozen importlib._bootstrap>:1204: in _gcd_import
    ???
<frozen importlib._bootstrap>:1176: in _find_and_load
    ???
<frozen importlib._bootstrap>:1147: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:690: in _load_unlocked
    ???
/Users/ME/.cache/pants/named_caches/pex_root/venvs/1/s/79b3222f/venv/lib/python3.11/site-packages/_pytest/assertion/rewrite.py:184: in exec_module
    exec(co, module.__dict__)
src/python/shoalsoft/FOO/BAR_test.py:17: in <module>
    from pants.engine.rules import QueryRule
/Users/ME/.cache/pants/named_caches/pex_root/venvs/1/s/79b3222f/venv/lib/python3.11/site-packages/pants/engine/rules.py:32: in <module>
    from pants.engine.engine_aware import SideEffecting
/Users/ME/.cache/pants/named_caches/pex_root/venvs/1/s/79b3222f/venv/lib/python3.11/site-packages/pants/engine/engine_aware.py:9: in <module>
    from pants.engine.internals import native_engine
E   ImportError: dlopen(/Users/ME/.cache/pants/named_caches/pex_root/venvs/1/s/79b3222f/venv/lib/python3.11/site-packages/pants/engine/internals/native_engine.so, 0x0002): tried: '/Users/ME/.cache/pants/named_caches/pex_root/venvs/1/s/79b3222f/venv/lib/python3.11/site-packages/pants/engine/internals/native_engine.so' (mach-o file, but is an incompatible architecture (have 'arm64', need 'x86_64h' or 'x86_64')), '/System/Volumes/Preboot/Cryptexes/OS/Users/tdyas/.cache/pants/named_caches/pex_root/venvs/1/s/79b3222f/venv/lib/python3.11/site-packages/pants/engine/internals/native_engine.so' (no such file), '/Users/tdyas/.cache/pants/named_caches/pex_root/venvs/1/s/79b3222f/venv/lib/python3.11/site-packages/pants/engine/internals/native_engine.so' (mach-o file, but is an incompatible architecture (have 'arm64', need 'x86_64h' or 'x86_64')), '/Users/tdyas/.cache/pants/named_caches/pex_root/venvs/1/9d4bc213af4e92b06fa120abcea29d2f03f0eeef/2982e4bb26bb389cf345ae6b2c01416ccd8d4126/lib/python3.11/site-packages/pants/engine/internals/native_engine.so' (mach-o file, but is an incompatible architecture (have 'arm64', need 'x86_64h' or 'x86_64')), '/System/Volumes/Preboot/Cryptexes/OS/Users/tdyas/.cache/pants/named_caches/pex_root/venvs/1/9d4bc213af4e92b06fa120abcea29d2f03f0eeef/2982e4bb26bb389cf345ae6b2c01416ccd8d4126/lib/python3.11/site-packages/pants/engine/internals/native_engine.so' (no such file), '/Users/tdyas/.cache/pants/named_caches/pex_root/venvs/1/9d4bc213af4e92b06fa120abcea29d2f03f0eeef/2982e4bb26bb389cf345ae6b2c01416ccd8d4126/lib/python3.11/site-packages/pants/engine/internals/native_engine.so' (mach-o file, but is an incompatible architecture (have 'arm64', need 'x86_64h' or 'x86_64'))
- generated xml file: src.python.shoalsoft.FOO.BAR_test.py.tests@@parametrize=pants-2.25.x.xml -
=========================== short test summary info ============================
ERROR src/python/shoalsoft/FOO/BAR_test.py
!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!
=============================== 1 error in 0.25s ===============================
```

--- a/.github/workflows/release.yaml
+++ b/.github/workflows/release.yaml
@@ -254,10 +254,12 @@ jobs:
         go-version: 1.19.5
     - env:
         ARCHFLAGS: -arch x86_64
+        _PYTHON_HOST_PLATFORM: macosx-13.0-x86_64
       name: Build wheels
       run: ./pants run src/python/pants_release/release.py -- build-wheels
     - env:
         ARCHFLAGS: -arch x86_64
+        _PYTHON_HOST_PLATFORM: macosx-13.0-x86_64
       name: Build Pants PEX
       run: ./pants package src/python/pants:pants-pex
     - continue-on-error: true
@@ -351,10 +353,12 @@ jobs:
         go-version: 1.19.5
     - env:
         ARCHFLAGS: -arch arm64
+        _PYTHON_HOST_PLATFORM: macosx-14.0-arm64
       name: Build wheels
       run: ./pants run src/python/pants_release/release.py -- build-wheels
     - env:
         ARCHFLAGS: -arch arm64
+        _PYTHON_HOST_PLATFORM: macosx-14.0-arm64
       name: Build Pants PEX
       run: ./pants package src/python/pants:pants-pex
     - continue-on-error: true

--- a/.github/workflows/test.yaml
+++ b/.github/workflows/test.yaml
@@ -532,10 +532,12 @@ jobs:
         go-version: 1.19.5
     - env:
         ARCHFLAGS: -arch x86_64
+        _PYTHON_HOST_PLATFORM: macosx-13.0-x86_64
       name: Build wheels
       run: ./pants run src/python/pants_release/release.py -- build-wheels
     - env:
         ARCHFLAGS: -arch x86_64
+        _PYTHON_HOST_PLATFORM: macosx-13.0-x86_64
       name: Build Pants PEX
       run: ./pants package src/python/pants:pants-pex
     - continue-on-error: true
@@ -612,10 +614,12 @@ jobs:
         go-version: 1.19.5
     - env:
         ARCHFLAGS: -arch arm64
+        _PYTHON_HOST_PLATFORM: macosx-14.0-arm64
       name: Build wheels
       run: ./pants run src/python/pants_release/release.py -- build-wheels
     - env:
         ARCHFLAGS: -arch arm64
+        _PYTHON_HOST_PLATFORM: macosx-14.0-arm64
       name: Build Pants PEX
       run: ./pants package src/python/pants:pants-pex
     - continue-on-error: true
@@ -1870,6 +1874,7 @@ jobs:
   test_python_macos13_x86_64:
     env:
       ARCHFLAGS: -arch x86_64
+      _PYTHON_HOST_PLATFORM: macosx-13.0-x86_64
     if: (github.repository_owner == 'pantsbuild') && (needs.classify_changes.outputs.docs_only != 'true')
     name: Test Python (macOS13-x86_64)
     needs:

--- a/src/python/pants/BUILD
+++ b/src/python/pants/BUILD
@@ -38,6 +38,10 @@ python_distribution(
         python_requires="==3.11.*",
     ),
     entry_points={"console_scripts": {"pants": "pants.bin.pants_loader:main"}},
+    # We need to explicitly control the wheel tagging, rather than following whatever the current
+    # Python interpreter is tagged for (especially on macOS, where a 'universal' interpreter build
+    # can lead to single-platform wheels being tagged as `universal2` wheels, incorrectly)
+    env_vars=["_PYTHON_HOST_PLATFORM"],
 )
 
 pex_binary(

--- a/src/python/pants_release/generate_github_workflows.py
+++ b/src/python/pants_release/generate_github_workflows.py
@@ -443,8 +443,17 @@ class Helper:
             # Works around bad `-arch arm64` flag embedded in Xcode 12.x Python interpreters on
             # intel machines. See: https://github.com/giampaolo/psutil/issues/1832
             ret["ARCHFLAGS"] = "-arch x86_64"
+            # Be explicit about the platform we've built our native code for: without this, Python
+            # builds and tags wheels based on the interpreter's build.
+            #
+            # At the time of writing, the GitHub-hosted runners' Pythons are built as
+            # macosx-10.9-universal2. Thus, without this env var, we tag our single-platform wheels
+            # as universal2... this is a lie and understandably leads to installing the wrong wheel
+            # and thus things do not work (see #21938 for an example).
+            ret["_PYTHON_HOST_PLATFORM"] = "macosx-13.0-x86_64"
         if self.platform in {Platform.MACOS14_ARM64}:
             ret["ARCHFLAGS"] = "-arch arm64"
+            ret["_PYTHON_HOST_PLATFORM"] = "macosx-14.0-arm64"
         if self.platform == Platform.LINUX_X86_64:
             # Currently we run Linux x86_64 CI on GitHub Actions-hosted hardware, and
             # these are weak dual-core machines. Default parallelism on those machines

--- a/src/python/pants_release/release.py
+++ b/src/python/pants_release/release.py
@@ -532,19 +532,20 @@ def build_pants_wheels() -> None:
 
     for package in PACKAGES:
         found_wheels = sorted(Path("dist").glob(f"{package}-{version}-*.whl"))
-        # NB: For any platform-specific wheels, like pantsbuild.pants, we assume that the
-        # top-level `dist` will only have wheels built for the current platform. This
-        # should be safe because it is not possible to build native wheels for another
-        # platform.
-        if not is_cross_platform(found_wheels) and len(found_wheels) > 1:
-            die(
-                softwrap(
-                    f"""
-                    Found multiple wheels for {package} in the `dist/` folder, but was
-                    expecting only one wheel: {sorted(wheel.name for wheel in found_wheels)}.
-                    """
+        if not is_cross_platform(found_wheels):
+            # NB: For any platform-specific wheels, like pantsbuild.pants, we assume that the
+            # top-level `dist` will only have wheels built for the current platform. This
+            # should be safe because it is not possible to build native wheels for another
+            # platform.
+            if len(found_wheels) > 1:
+                die(
+                    softwrap(
+                        f"""
+                        Found multiple wheels for {package} in the `dist/` folder, but was
+                        expecting only one wheel: {sorted(wheel.name for wheel in found_wheels)}.
+                        """
+                    )
                 )
-            )
         for wheel in found_wheels:
             wheel_dest = dest / wheel.name
             if not wheel_dest.exists():

--- a/src/python/pants_release/release.py
+++ b/src/python/pants_release/release.py
@@ -546,6 +546,21 @@ def build_pants_wheels() -> None:
                         """
                     )
                 )
+
+            # We also only build for a single architecture at a time, so lets confirm that the wheel
+            # isn't potentially reporting itself as applicable to arm64 and x86-64 ('universal2', in
+            # macOS parlance) (see #21938):
+            wheel = found_wheels[0]
+            if "universal2" in str(wheel):
+                die(
+                    softwrap(
+                        f"""
+                        Found universal wheel for {package} in the `dist/` folder, but was
+                        expecting a specific architecture: {wheel}.
+                        """
+                    )
+                )
+
         for wheel in found_wheels:
             wheel_dest = dest / wheel.name
             if not wheel_dest.exists():

